examples_tests/23.Autoexposure/main.cpp

```diff
@@ -4,20 +4,771 @@
+				asset::ISampler::ETC_CLAMP_TO_EDGE,
+				asset::ISampler::ETBC_FLOAT_OPAQUE_BLACK,
+				asset::ISampler::ETF_NEAREST,
+				asset::ISampler::ETF_NEAREST,
+				asset::ISampler::ESMM_LINEAR,
+				0u,
+				false,
+				nbl::asset::ECO_ALWAYS
+			};
+			core::smart_refctd_ptr<video::IGPUSampler> sampler = logicalDevice->createGPUSampler(std::move(samplerParams));
+			video::IGPUDescriptorSetLayout::SBinding binding =
+			{
+				0u,
+				nbl::asset::EDT_COMBINED_IMAGE_SAMPLER,
+				1u,
+				nbl::video::IGPUShader::ESS_FRAGMENT,
+				&sampler
+			};
+
+			fstDSLayout = logicalDevice->createGPUDescriptorSetLayout(&binding, &binding + 1u);
+			fstPipelineLayout = logicalDevice->createGPUPipelineLayout(nullptr, nullptr, nullptr, nullptr, nullptr, core::smart_refctd_ptr(fstDSLayout));
+		}
+
+		// create FST renderpass independent pipeline
+		auto fstProtoPipeline = nbl::ext::FullScreenTriangle::createProtoPipeline(cpu2gpuParams);
+		auto fstRenderpassIndep = ext::FullScreenTriangle::createRenderpassIndependentPipeline(
+			logicalDevice.get(),
+			fstProtoPipeline,
+			std::move(fstFragShaderGPU),
+			std::move(fstPipelineLayout));
+
+		// create & update FST DS
+		{
+			const uint32_t setCount = SC_IMG_COUNT;
+			auto dsPool = logicalDevice->createDescriptorPoolForDSLayouts(
+				video::IDescriptorPool::ECF_NONE,
+				&fstDSLayout.get(), &fstDSLayout.get() + 1ull,
+				&setCount);
+
+			for (uint32_t i = 0u; i < SC_IMG_COUNT; ++i)
+			{
+				fstDS[i] = logicalDevice->createGPUDescriptorSet(dsPool.get(), core::smart_refctd_ptr(fstDSLayout));
+
+				video::IGPUDescriptorSet::SWriteDescriptorSet write = {};
+				write.dstSet = fstDS[i].get();
+				write.binding = 0u;
+				write.arrayElement = 0u;
+				write.count = 1u;
+				write.descriptorType = asset::EDT_COMBINED_IMAGE_SAMPLER;
+
+				video::IGPUDescriptorSet::SDescriptorInfo info = {};
+				info.desc = tonemapperOutImageViews[i];
+				info.image.sampler = nullptr;
+				info.image.imageLayout = asset::EIL_SHADER_READ_ONLY_OPTIMAL;
+
+				write.info = &info;
+
+				logicalDevice->updateDescriptorSets(1u, &write, 0u, nullptr);
+
+			}
+		}
+
+		// create FST graphics pipline
+		{
+			nbl::video::IGPUGraphicsPipeline::SCreationParams creationParams = {};
+			creationParams.renderpassIndependent = core::smart_refctd_ptr<nbl::video::IGPURenderpassIndependentPipeline>(const_cast<video::IGPURenderpassIndependentPipeline*>(fstRenderpassIndep.get()));
+			creationParams.renderpass = core::smart_refctd_ptr(renderpass);
+
+			fstGraphicsPipeline = logicalDevice->createGPUGraphicsPipeline(nullptr, std::move(creationParams));
+		}
+
+		for (uint32_t i = 0u; i < FRAMES_IN_FLIGHT; i++)
+		{
+			imageAcquire[i] = logicalDevice->createSemaphore();
+			renderFinished[i] = logicalDevice->createSemaphore();
+		}
+
+		lastPresentStamp = std::chrono::high_resolution_clock::now();
+	}
+
+	void onAppTerminated_impl() override
+	{
+		logicalDevice->waitIdle();
+	}
+
+	void workLoopBody() override
```

Comment:
would appreciate (if possible `static`) methods in the extension classes to "suggest" barriers (or just to remove a bit of verbosity from the struct setups